### ISSUE TITLE
Multiple Jep instances can slow down throughput/concurrency waiting on CPython GIL

### ISSUE THREAD

**Issue by wizbots** [ninia/jep#34]
Hello. Thanks for creating JEP! 

We are using it to integrate a python based text analytics library we have developed that under the covers uses sklearn, scipy and numpy. It works great with a single instance running on main thread. It is performant and gives correct results.

However, we are now testing using it with Multiple Threads. The test allocates a set of threads, and each thread allocates its own JEP instance and executes a number of calls. Each call invokes a function, and we are not setting any variables or anything. Call looks like this:

``` java
ArrayList<Map<String,Object>> results = (ArrayList<Map<String,Object>>) jep.getValue('extract_data("This is a string")');
```

It mostly works but for 2 issues:
1. The more threads I create (on a quad core Mac Server) the slower the throughput.
2. I am getting an exception in the numpy core. I will spare you the details of the stack trace there since it is very particular to our code base.

While the second issue is troubling - the first one is really enigmatic. Can you shed any light on this. In my measurements I am careful to separate out the load times for the library and prime it with a function call before using it with timing.

Also, I notice that I cannot create a Pool of JEP instances and use them by multiple threads - much as I would a connection pool for a database. Is that correct?

Many thanks,

Conor

**Comment by wizbots** [comment]
As an extra data point - my second issue went away when I did not invoke close when cleaning up my thread....

**Comment by ndjensen** [comment]
Hi!  Hopefully I can shed some insight.

1. I suspect you're running into the GIL.  Have you read this page on [Jep and the GIL](https://github.com/mrj0/jep/wiki/Jep-and-the-GIL)?  See if you think that applies to your scenario.
2. This sounds like a problem with numpy we've been seeing.  Check out #28 which has quite a bit of discussion and some related tickets.  We know there's a problem with numpy's `any()` and `str()` methods, if you're having a problem with a different method you should add a comment to #28 with your particular stacktrace.  (Or you can also just add a comment such as "I hit this too" so we better recognize how widespread the problem is...)

I'm not sure I follow the problem with pools?  You should be able to create a pool, but unfortunately it's essential that the same thread that creates the interpreter is the one using the interpreter.  Therefore if you do make a pool you need to key off the thread for reusing an interpreter (vs making a new one), or else make an ExecutorService.

**Comment by wizbots** [comment]
Hi. Thanks for the response. I commented on #28. 

I had read that wiki page, but I misunderstood the GIL implications. The scenario certainly seems to apply. 

Is it the case that there is a single GIL for all Jep instances then? I was assuming they would be loading their own instances of python. But if we have a single lock for all Jep instances it would explain what I am seeing - which is roughly a 30% reduction in throughput running 5 threads vs 1 thread, for the same workload.

**Comment by ndjensen** [comment]
Yeah, the GIL is for the entire process unfortunately.  So the more Jep instances, the more likely the throughput will degrade unless the various Jep instances are releasing the GIL as described in the wiki page.

Except for the workarounds mentioned in the wiki, we don't have a good solution around the GIL at this time.  I am hoping that with more developers using Jep we will either get enough creative minds to come up with a better solution or attract the attention of the CPython developers who will then put more effort into resolving the limitations of the GIL.

**Comment by ndjensen** [comment]
I updated the wiki a bit to try and make it more clear.  @bsteffensmeier is dreaming up solutions to the numpy problem, but any potential solutions will have to pass numpy tests and code review.  If they got through that, we'd still have to wait for a future release of numpy.

I'd like to keep this ticket open as reference for others, though I don't know at this time how we can fix it.  For the foreseeable future it seems to be a limitation of CPython and we will continue to encourage [the workarounds mentioned in the wiki](https://github.com/mrj0/jep/wiki/Jep-and-the-GIL#how-can-i-work-around-gil-performance-issues) until we come up with something better.

**Comment by wizbots** [comment]
Thanks. By the way, my work around is to queue requests from calling threads in a blocking queue, and suspend the caller, then have a single thread serve that queue and resume the caller when each task is done. So basically single threaded execution with no meaningful additional overhead, and allowing multi-threaded invocation from clients.

**Comment by Techcable** [comment]
Wouldn't it be possible to have each jep instance execute the python code in its own subprocess?
Wouldn't the overhead of those processes be far less than that of a highly contented GIL?
This could scale up to 1000s of jep instances on linux.

**Comment by bsteffensmeier** [comment]
It's just a matter of project scope. Jep is designed to support quick interoperability between python and java environments. If separate processes are started you lose the ability to call back into the original JVM. Passing data back and forth would require defining a custom mechanism of interprocess communication and remote method invocation. While this is feasible and there are use cases where it would be better than what jep does now, there is almost no overlap between the existing jep code and what would be needed to implement subprocesses efficiently so I don't think it makes sense to do it within the current scope of jep.

Jep has been successfully used with the multiprocessing module in python to spawn off child processes. I'm not sure how well this performs and it is limited to passing pure python objects using the multprocessing API, but it might be an area to look into. You could set it up to use jep to send the data between python and java and then use the multiprocessing API to send the hard work to another process. Theoretically you could get better performance if you cut jep out and send information directly between java and python in separate processes because you would avoid the intermediary representation in jep. This direct communication seems to be what is provided by the (Py4j)[https://www.py4j.org/] project.

**Comment by Techcable** [comment]
I forgot about inter-process communication.
It Makes sense to let the Py4j project handle it.
@bsteffensmeier 
You switched the () and [] on the markdown link.

**Comment by pradeepvb** [comment]
Is it possible to use JEP with multiprocessing without having the subprocess use the spawn and be listed as regular python process or a java process without using the parent's process arguments.

**Comment by ndjensen** [comment]
I'm not sure I follow the question.  With multiprocessing you change the process name when [constructing the object](https://docs.python.org/2/library/multiprocessing.html#multiprocessing.Process) or [with a pool](http://stackoverflow.com/questions/26923050/how-do-i-name-the-processes-in-a-multiprocessing-pool).  I don't think that changes how the subprocess appears in a tool like top.

I have not used [exec methods](https://docs.python.org/2/library/os.html#process-management).  I'm not sure what would happen if you used those within a subprocess.

Multiprocessing with Jep needs more experimentation.  Please share any findings in case others have the same questions.

**Comment by bsteffensmeier** [comment]
There is some movement within Python to make the GIL [a per interpreter lock](https://www.blog.pythonlibrary.org/2018/04/02/pydev-of-the-week-eric-snow/) which would allow sub-interpreters to run with separate locks. These changes are starting with [PEP-554,](https://www.python.org/dev/peps/pep-0554/) so I would recommend anyone who wants this functionality should support that PEP. I do not think we can ever resolve this issue without these types of changes in Python.

**Comment by ndjensen** [comment]
This is a really old ticket.  It appears it was written against Jep 3.3 when I added numpy support.  For the record, newer versions of Jep release the GIL **every** time Python calls back into Java, so the problem should be alleviated a bit in newer versions.  That said, for long running Python code, I still don't know how to overcome GIL limitations other than what we've already done.

**Comment by kaiseu** [comment]
do we have some suggestions on using JEP with multiply threads concurrently, for my scenario, I created one static JEP instance but when call in multi thread it failed with "jep.JepException: Invalid thread access". But when I try with multi JEP instance, it failed with "Jep.setJepLibraryPath(...) called after initializing python interpreter".

**Comment by ndjensen** [comment]
@kaiseu, you can use multiple Java threads with Jep.  Just use a separate Jep instance for each Java thread.  You can only set the Jep library path once and it should be set before you instantiate any Jep threads.

**Comment by kaiseu** [comment]
@ndjensen Thanks for the info, need I set the Jep library path every time before  I instantiate JEP instance if it's running distributedly, by distributedly I mean something like Spark map reduce job, in which each task calls a python function.

**Comment by kaiseu** [comment]
when trying to load jep.so, I got the error: undefined symbol: Py_HashRandomizationFlag, what's wrong with my .so file?

**Comment by cowwoc** [comment]
@ndjensen FYI, JDK 16 adds https://openjdk.java.net/jeps/380 across all major platforms. This allows you to implement extremely efficient interprocess communication.

I think there is strong value in moving this way because not all python modules support multiple sub-interpreters. I think that jpype uses multiple processes as well and between the two of you that's the majority of the python-java bridge market share.

**Comment by bsteffensmeier** [comment]
With Jep 4.2 and Python 3.12 it is now possible to create sub-interpreters that do not share a GIL and offer more complete concurrency so I am closing this issue. See the [Jep 4.2 release notes](https://github.com/ninia/jep/blob/master/release_notes/4.2-notes.rst#support-for-isolated-subinterpreter) for more details
